board.rb

```diff
@@ -0,0 +1,324 @@
+# If *not* in an array, can directly access.
+      if row_i == WIDTH-2 || row_i == WIDTH-1
+        row.map do |square|
+          square.first.color = "white"
+        end
+      end
+    end
+    @board
+  end
+
+  #  def remove
+  #   #TODO: remove pieces if piece.captured == true
+  #   #stores removed pieces in removed array.
+  #   @board.map.with_index |row, row_i|
+  #     row.map.with_index
+  #   end
+  # end
+
+  def move_valid(coordinate)
+    @poss_moves.clear
+    square = @board[8-coordinate[1]][LETTER[coordinate[0]]]
+    if square.empty?
+      puts "This is empty!"
+    else
+      piece = square.first
+    end
+    p piece
+    # piece.moves.each do |coordinate|
+    # @board.each_with_index do |row, row_i|
+    #   row.each_with_index do |square, square_i|
+    piece.moves.each do |coordinate_set|
+      coordinate_set.each do |x,y|
+        if !@board[x][y].empty?
+          if piece.color != @board[x][y][0].color
+            @poss_moves << [x, y]
+            break
+          else
+            break
+          end
+        end
+        @poss_moves 5
+        << [x, y] if @board[x][y].empty?
+      end
+    end
+  end
+
+  # def make_move(piece)
+  #   piece => find possible moves (give coordinates to user)
+  #   they will give us coordinate they want to move to
+  #   => place piece (piece, coordinate)
+  #   move the piece to cooresponding coordinate
+  def place_piece(piece, new_square)
+
+
+  end
+
+  def translate(input)
+
+  end
+
+  def remove
+
+  end
+
+  def to_s
+    @board.each_with_index.map do |row, i|
+      # puts "#{8-i} #{row}"
+      print "#{8-i}  "
+      row.each_with_index do |square, square_i|
+        if square.empty?
+          print "_ "
+          next
+        end
+        print "#{square.first.class.to_s == "Knight" ? "N" : square.first.class.to_s[0]} " if square.first.color == "black"
+        print "#{square.first.class.to_s == "Knight" ? "n" : square.first.class.to_s[0].downcase} " if square.first.color == "white"
+      end
+      puts
+    end#.join(" \n")
+    "   " + [*"a".."h"].join(" ")
+  end
+end
+
+
+class Piece
+  attr_reader :captured
+  attr_accessor :color
+
+  def initialize(arguments, color = "black", captured = false)
+    @captured, @color = captured, color
+    @x, @y = arguments[0], arguments[1]
+    @moves = []
+  end
+
+  def remove_nil
+    @moves.each do |x|
+      @moves.delete(x) if x.empty?
+    end
+  end
+
+  def captured!
+    @captured = !@captured
+  end
+
+  def to_s
+    "#{@color.upcase} #{(self.color == "black" ? self.name : self.name.downcase)} moves: #{@moves.flatten(1)}"
+  end
+end
+
+class King < Piece
+  attr_reader :name
+  def initialize(arguments)
+    super(arguments)
+    @name = "KING"
+  end
+
+  def moves
+    [*-1..1].permutation(2).to_a.each do |dx,dy|
+      next if dx == 0 && dy == 0
+      @moves << [[@x+dx, @y+dy]] unless @x+dx > WIDTH-1 || @x+dx < 0 || @y+dy > WIDTH-1 || @y+dy < 0
```

Comment:
Seems like the bounds checking code gets used a lot, maybe we could make that a method?